### PR TITLE
New Vagrant file for creating PXEBOOT server

### PR DESCRIPTION
**Pull request by jirihnidek**
* New Vagrant file is in new directory anaconda, because it was
  the simpliest way to not trigger other ansible scripts
* It downloads image of Centos7 minimal image
* I have plan to modify Ansible scripts to support Fedora 27
  in the future
* The IP address of pxe server has to be static, because it is
  necessary to know it before VM with PXE server is created.
  The is used, when private network is created. Otherwise
  client will not boot using PXE.
* The client for testing machine can be created simply in
  virt-manager, but it will requires at least 2GB RAM due to
  some bug in anaconda.

### CODE REVIEW

**Comment by jirihnidek** [comment]
[[Installer] create a PXEBOOT server for centos7 install via vagrant](https://trello.com/c/QYpLZt3I/777-installer-create-a-pxeboot-server-for-centos7-install-via-vagrant)

File: vagrant/roles/pxe-server/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    state: directory
+    owner: vagrant
+    group: vagrant
+    mode: 0755
+  become: true
+
+- name: check if installation iso image is already downloaded
+  stat:
+    path: "/vagrant_data/{{iso_img_file}}"
+  register: iso_img_stat_result
```
Comment (kahowell) [inline vagrant/roles/pxe-server/tasks/main.yml]: This is unnecessary as get_url will not re-download the file :smile:

File: vagrant/roles/pxe-server/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+- name: download minimal installation iso image
+  get_url:
+    url: "{{iso_img_url}}"
+    dest: /vagrant_data
+    checksum: "{{iso_img_checksum}}"
+  when: iso_img_stat_result.stat.exists == False
+
+- name: mount iso image
```
Comment (kahowell) [inline vagrant/roles/pxe-server/tasks/main.yml]: Instead of mount, copy, unmount, we may be able to simply use `iso_extract` (http://docs.ansible.com/ansible/latest/iso_extract_module.html), only downside is that it requires a list of files to copy, but we might not need all the files anyway. Please try it and let me know if it's a viable option.

File: anaconda/Vagrantfile [hunk 3]
```diff
@@ -0,0 +1,30 @@
+# We need to specify static IP address, because this IP address has to be part
```
[the remark this answers is not in the document]
Comment (jirihnidek) [reply anaconda/Vagrantfile]: I agree. I wasn't happy with separate Vagrant file. Adding existing host into new group sounds good.

File: anaconda/Vagrantfile [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+  config.vm.provision "ansible", run: "always" do |ansible|
+    ansible.playbook = "../vagrant/vagrant_pxe_server.yml"
+    ansible.groups = {
+      "pxe-servers" => [
+        "pxe-server"
+      ]
+    }
+    ansible.extra_vars = {
+      "pxe_server_ip_addr" => ip_addr + '/' + ip_prefix
```
[the remark this answers is not in the document]
Comment (jirihnidek) [reply anaconda/Vagrantfile]: I'm sorry. I don't understand. The value is the same as what?

File: vagrant/roles/pxe-server/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    group: vagrant
+    mode: 0755
+  become: true
+
+- name: check if installation iso image is already downloaded
+  stat:
+    path: "/vagrant_data/{{iso_img_file}}"
+  register: iso_img_stat_result
+
+- name: download minimal installation iso image
```
[the remark this answers is not in the document]
Comment (jirihnidek) [reply vagrant/roles/pxe-server/tasks/main.yml]: Good point. I will try `delegate_to: localhost` and `local_action`.

File: vagrant/roles/pxe-server/tasks/main.yml (hunk 2 above)
Comment (jirihnidek) [reply vagrant/roles/pxe-server/tasks/main.yml]: I will try it.

File: vagrant/roles/pxe-server/tasks/main.yml (hunk 1 above)
Comment (jirihnidek) [reply vagrant/roles/pxe-server/tasks/main.yml]: Sure :smiley:

File: vagrant/roles/pxe-server/tasks/main.yml (hunk 2 above)
Comment (jirihnidek) [reply vagrant/roles/pxe-server/tasks/main.yml]: I think it will not work for us, because you have to explicitly say what files you want to extract from ISO and you cannot extract directories. :disappointed: We IMHO need almost all files from ISO and it would be hard to maintain such list. The status of this module is preview. Thus functionality of `iso_extract` can be changed in the future.

File: vagrant/roles/pxe-server/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+- name: start vsftpd service
+  service:
+    name: vsftpd
+    state: started
+  become: true
+
+- name: download minimal installation iso image to localhost
+  get_url:
+    url: "{{iso_img_url}}"
+    dest: vagrant_data
```
Comment (kahowell) [inline vagrant/roles/pxe-server/tasks/main.yml]: Please use the path to the actual file, not the parent dir, otherwise the download still happens! (Did not know this prior).

This worked for me: `dest: "vagrant_data/{{iso_img_file}}"`

File: vagrant/roles/pxe-server/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    path: /vagrant_data
+    state: directory
+    owner: vagrant
+    group: vagrant
+    mode: 0755
+  become: true
+
+- name: copy iso image to pxe server
+  copy:
+    src: vagrant_data
```
Comment (kahowell) [inline vagrant/roles/pxe-server/tasks/main.yml]: Needs to end in `/`; otherwise, we end up with `/vagrant_data/vagrant_data/*.iso` instead of `/vagrant_data/*.iso`

File: .gitignore [hunk 8]
```diff
@@ -33,3 +33,4 @@ dist
 .settings
 *.retry
 .vagrant
+vagrant/vagrant_data/*.iso
```
Comment (kahowell) [inline .gitignore]: Please add vagrant/vagrant_data to the Vagrantfile rsync excludes... we don't want to copy this file indiscriminately.

File: vagrant/vagrant.yml [hunk 9]
```diff
@@ -1,15 +1,15 @@
 ---
-- hosts: "*fedora*"
+- hosts: "*fedora*, subman-devel"
```
Comment (kahowell) [inline vagrant/vagrant.yml]: Unnecessary change, and might actually cause issues as written. Just don't change this line.

File: vagrant/roles/pxe-server/tasks/main.yml [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+    dest: /etc/dhcp/dhcpd.conf
+  become: yes
+
+- name: enable dhcpd service
+  service:
+    name: dhcpd
+    enabled: yes
+  become: true
+
+- name: start dhcpd service
```
[the remark this answers is not in the document]
Comment (jirihnidek) [reply vagrant/roles/pxe-server/tasks/main.yml]: Thanks. I didn't know it. I will combine it into one statement.

File: vagrant/roles/pxe-server/defaults/main.yml [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+  - dhcp
+  - tftp
+  - tftp-server
+  - syslinux
+  - wget
+  - vsftpd
+
+# Newer version is available e.g. here: http://ftp.cvut.cz/centos/7/isos/x86_64/
+iso_img_server: "http://ftp.cvut.cz/centos/7/isos/x86_64"
+iso_img_file: "CentOS-7-x86_64-Minimal-1708.iso"
```
[the remark this answers is not in the document]
Comment (jirihnidek) [reply vagrant/roles/pxe-server/defaults/main.yml]: CVUT wasn't my choice :-) (Anaconda did this). I will change centos7 iso to Fedora and URL will generic. The fastest mirror will be found during HTTP request.

File: Vagrantfile [hunk 12]
```diff
@@ -94,3 +97,34 @@ Vagrant.configure("2") do |config|
...
+
+# We need to specify static IP address, because this IP address has to be part
+# of private network configuration, see: libvirt__dhcp_bootp_server: ip_addr
+ip_addr = '192.168.111.5'
+ip_prefix = '24'
+
+Vagrant.configure("2") do |config|
+  config.vm.hostname = 'pxe-server'
+  config.vm.box = 'centos/7'
+  config.vm.define 'pxe-server'
```
[the remark this answers is not in the document]
Comment (jirihnidek) [reply Vagrantfile]: Good point. Thanks.

File: Vagrantfile [hunk 13]
```diff
@@ -94,3 +97,34 @@ Vagrant.configure("2") do |config|
...
+  config.vm.provision "ansible", run: "always" do |ansible|
+    ansible.playbook = "vagrant/vagrant_pxe_server.yml"
+    ansible.groups = {
+      "pxe-servers" => [
+        "pxe-server"
+      ]
+    }
+    ansible.extra_vars = {
+      "pxe_server_ip_addr" => ip_addr + '/' + ip_prefix
+    }
```
[the remark this answers is not in the document]
Comment (jirihnidek) [reply Vagrantfile]: I understand now, but I don't agree :smiley:. Imagine following situation. You already have subnet `192.168.155.0` (e.g. for some network used by VMs) and you have to use some different subnet for private network `pxe-server0`, then it is possible to change it only in `Vagrantfile` (due to definition of private network) and you need to pass this different value into Ansible playbook using `ansible.extra_vars`.

The value in `defaults/main.yml` is used only in situation, when you e.g. run Ansible playbook manually. I can add note about it to `defaults/main.yml`.

File: vagrant/roles/pxe-server/tasks/main.yml (hunk 6 above)
Comment (jirihnidek) [reply vagrant/roles/pxe-server/tasks/main.yml]: I use local web server for testing and I didn't notice it. Thanks.

File: vagrant/roles/pxe-server/tasks/main.yml (hunk 7 above)
Comment (jirihnidek) [reply vagrant/roles/pxe-server/tasks/main.yml]: Yeah. I already noticed it.

File: vagrant/vagrant.yml (hunk 9 above)
Comment (jirihnidek) [reply vagrant/vagrant.yml]: OK

File: Vagrantfile (hunk 13 above)
[the remark this answers is not in the document]
Comment (kahowell) [reply Vagrantfile]: Makes sense. Thanks for the explanation.

File: vagrant/roles/pxe-server/defaults/main.yml [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+# iso_img_file: "CentOS-7-x86_64-Minimal-1708.iso"
+# # Fresh sha256 checksums of CentOS images can be found e.g. here:
+# # http://ftp.cvut.cz/centos/7/isos/x86_64/sha256sum.txt
+# iso_img_checksum: "sha256:bba314624956961a2ea31dd460cd860a77911c1e0a56e4820a12b9c5dad363f5"
+
+##### Fedora 27 #####
+
+distro_name: "fedora"
+distro_version: "27"
+# Newer version is available e.g. here: http://ftp.cvut.cz/centos/7/isos/x86_64/
```
Comment (kahowell) [inline vagrant/roles/pxe-server/defaults/main.yml]: nitpick: comment needs different url

**Comment by jirihnidek** [comment]
Squashed and ready for merge.
